### PR TITLE
Resolve ESIMD sort KT compilation failures with new compiler

### PR DESCRIPTION
**Pull request by mmichel11**
Our ESIMD radix sort has two issues with newer compilers:

1. We unnecessarily use the `template` keyword prior to calling a function without a template argument list. This causes compilation failures with newer compilers. This has been fixed by removing `template` in this instance.
2. The `sycl::ext::intel::esimd::sw_barrier` enumerator has been removed and as far as I am aware there is no other ESIMD equivalent. We have a note that these `sw_barrier` fences are for performance reasons. There is one fence in the inter-wg synchronization that I believe is necessary for correctness (I see a hang otherwise). However, it should use a global memory fence and not a software barrier as we need ordering of reads which are from global memory. I have removed the other sw_barrier fences and do not observe any measurable performance impact.

I can share performance data for 2 if requested.

### CODE REVIEW

**Review by danhoeflinger — DISMISSED** [review]
LGTM, 
I think its worthwhile to get @dmitriy-sobolev  approval before merging, unless there is urgent reason to go forward before next week.

**Comment by adamfidel** [comment]
I had to use the [Wayback Machine](https://web.archive.org/web/20230929210820/https://intel.github.io/llvm-docs/doxygen/group__sycl__esimd__memory.html) to see what `sw_barrier` was meant to do since it was scrubbed from the documentation after being removed.

Its intended effect was 

| Name| Description |
| ---|---|
| `sw_barrier` | Creates a software (compiler) barrier, which does not generate any instruction and only prevents instruction scheduler from reordering instructions across this barrier at compile time. |

It doesn't appear that that the fences that are being removed are guarding against any critical reordering across the fence from what I can tell, so I think it's fine there. It also looks like we're not touching any fences that were recently added in #1676.

It seems okay at a glance but I would still perhaps wait on input from @dmitriy-sobolev as suggested by @danhoeflinger.

**Comment by mmichel11** [comment]
I will wait on @dmitriy-sobolev's input next week before merging.

File: include/oneapi/dpl/experimental/kt/internal/esimd_radix_sort_kernels.h
Comment (dmitriy-sobolev) [inline include/oneapi/dpl/experimental/kt/internal/esimd_radix_sort_kernels.h:642]: Let me think about it since I do not have clear understanding of why there is a hang after removal of `sw_barrier` and why the introduced barrier helps... One confusing fact is that the fence is applied on group-scope by default (i.e. it guarantees global memory ops visibility within a group), however it is a cross-group synchronization section.

I was able to reproduce the hang while removing the fence. However, it does not reproduce when I replace it with:
`__dpl_esimd::__ns::fence<__dpl_esimd::__ns::memory_kind::local>();`

Note that this fence above:
https://github.com/oneapi-src/oneDPL/pull/1790/files#diff-b0f1146f4a5f7425f04ba4973e688c2c834dc1c4f7563ecb95a3e937bab81c1bR601-R603
guarantees that the global histogram and histograms local for each group are observable by other groups. This should be enough to avoid hanging. In the worst case, a group "N" can scan all the previous histograms.

Comment (dmitriy-sobolev) [reply include/oneapi/dpl/experimental/kt/internal/esimd_radix_sort_kernels.h:642]: I've gathered the kernels compiled in asm format for Intel® Data Center GPU Max:
 - [sw_fence.txt](https://github.com/user-attachments/files/16748126/sw_fence.txt)
 - [local_fence.txt](https://github.com/user-attachments/files/16748127/local_fence.txt)
 - [no_fence.txt](https://github.com/user-attachments/files/16748128/no_fence.txt)

The major difference between `sw_fence` and `no_fence` is the following:

```asm
_L_k43_10_preHeader:
BB_9:
// B018: Preds:{B017, B018},  Succs:{B018}
(W)     jmpi                                 _L_k43_10_preHeader                                     // $487
```
It looks like an plain endless loop.

I suggest changing the barrier to the local one, guarding it with a macro to turn it off after the next minor version of icpx, and adding a comment. I will create a bug ticket in the meantime.

Comment (mmichel11) [reply include/oneapi/dpl/experimental/kt/internal/esimd_radix_sort_kernels.h]: Thanks for taking a deeper look and clarifying on the barriers. I agree that it should not be needed.

I have switched to the local barrier, added a compiler check macro, and added a comment.

File: include/oneapi/dpl/experimental/kt/internal/esimd_defs.h
Comment (dmitriy-sobolev) [inline include/oneapi/dpl/experimental/kt/internal/esimd_defs.h:24]: Let's also apply the fix with [oneAPI DPC++ compiler](https://github.com/intel/llvm/tree/sycl?tab=readme-ov-file#oneapi-dpc-compiler): I was able to reproduce the hang with this compiler.

I would recommend relying on `_ONEDPL_LIBSYCL_VERSION` macro instead of `__INTEL_LLVM_COMPILER`. The drawback is that is not directly connected to the compiler version, which has the bug, but they are shipped side-by-side. I think it would be safe to assume `8` as a major version and `1` is a minor. The latest combination is 8.0, but 8.1 should provide more safety in case if a minor update happens soon.

Comment (dmitriy-sobolev) [reply include/oneapi/dpl/experimental/kt/internal/esimd_defs.h:24]: Probably, 9.0 would be even better.

Comment (dmitriy-sobolev) [reply include/oneapi/dpl/experimental/kt/internal/esimd_defs.h:24]: One more thing: there is only one place where the fix is applicable. You may consider moving the macro definition closer the the usage or even add the condition "in-place".

Comment (mmichel11) [reply include/oneapi/dpl/experimental/kt/internal/esimd_defs.h]: I've switched the macro to rely on `_ONEDPL_LIBSYCL_VERSION < 90000`, so it will expire with the next major update to 9. The comment above the macro explains the issue, so I have in-lined the check instead of making a separate macro.

**Review by dmitriy-sobolev — APPROVED** [review]
Thank you very much for the fix! LGTM.

**Review by danhoeflinger — APPROVED** [review]

**Pull request merged**
